extractors/src/main/boilerplate/sqlest/extractor/Extractor.scala.template

```diff
@@ -132,6 +167,32 @@ case class MappedExtractor[Row, A, B](inner: Extractor[Row, A], func: A => B, un
   def emit(accumulator: inner.Accumulator) = inner.emit(accumulator).map(func)
 }
 
+case class ChoiceExtractor[Row, A, B <: D, C <: D, D](inner: Extractor[Row, A], pred: A => Boolean, left: Extractor[Row, B], right: Extractor[Row, C]) extends Extractor[Row, D] with SimpleExtractor[Row, D] with SingleRowExtractor[Row, D] {
+  type Accumulator = Option[(inner.Accumulator, Option[D])]
+
+  private def runChoice(a: A, row: Row): Option[D] = if (pred(a)) runLeft(row) else runRight(row)
+  private def runLeft(row: Row): Option[D] = left.emit(left.initialize(row))
+  private def runRight(row: Row): Option[D] = right.emit(right.initialize(row))
+
+  def initialize(row: Row) = {
+    val innerAcc = inner.initialize(row)
```

Comment:
_.initialize(_.emit(row)) gets called on the `inner` extractor by every extractor in the chain in `cond` and `switch`. It would be nice if there was a way to structure this to avoid that - do you have any thoughts?